File: src/hymt/exec_wrapper.py
from __future__ import annotations

from collections.abc import Iterable
from dataclasses import dataclass
from pathlib import Path
import asyncio
import fnmatch
import json
import os
import subprocess
import sys
import threading
from typing import BinaryIO, TextIO

from hymt.config import HotConfig
from hymt.exec_cache import translate_cached_text


ANSI_CYAN = "\033[36m"
ANSI_YELLOW = "\033[33m"
ANSI_RESET = "\033[0m"


@dataclass(frozen=True)
class ExecResult:
    returncode: int
    stdout: bytes
    stderr: bytes


def run_exec_command(
    command: list[str], target_lang: str, config: HotConfig | None = None
) -> int:
    if not command:
        raise ValueError("command is required after '--'")
    active_config = config or HotConfig()
    result = _run_command(command)
    try:
        asyncio.run(_translate_result(command, result, target_lang, active_config))
    except KeyboardInterrupt:
        print("hymt: translation cancelled.", file=sys.stderr)
    except (OSError, RuntimeError, ValueError) as exc:
        print(f"hymt: translation failed: {exc}", file=sys.stderr)
    return result.returncode


def _run_command(command: list[str]) -> ExecResult:
    try:
        process = subprocess.Popen(
            command,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            close_fds=True,
        )
    except FileNotFoundError as exc:
        raise ValueError(f"Command not found: {command[0]}") from exc
    assert process.stdout is not None
    assert process.stderr is not None
    stdout_chunks: list[bytes] = []
    stderr_chunks: list[bytes] = []
    stdout_thread = threading.Thread(
        target=_copy_pipe,
        args=(process.stdout, sys.stdout, stdout_chunks),
        daemon=True,
    )
    stderr_thread = threading.Thread(
        target=_copy_pipe,
        args=(process.stderr, sys.stderr, stderr_chunks),
        daemon=True,
    )
    stdout_thread.start()
    stderr_thread.start()
    returncode = process.wait()
    stdout_thread.join()
    stderr_thread.join()
    return ExecResult(returncode, b"".join(stdout_chunks), b"".join(stderr_chunks))


def _copy_pipe(pipe: BinaryIO, stream: TextIO, chunks: list[bytes]) -> None:
    while True:
        chunk = pipe.read(8192)
        if not chunk:
            return
        chunks.append(chunk)
        _write_bytes(stream, chunk)


def _write_bytes(stream: TextIO, data: bytes) -> None:
    buffer = getattr(stream, "buffer", None)
    if buffer is not None:
        buffer.write(data)
        buffer.flush()
        return
    stream.write(data.decode("utf-8", errors="replace"))
    stream.flush()


async def _translate_result(
    command: list[str], result: ExecResult, target_lang: str, config: HotConfig
) -> None:
    if config.exec_translate_stderr and result.stderr:
        stderr_text = _decode_for_translation(result.stderr)
        if stderr_text is not None:
            translated = await _translate_output(
                command, stderr_text, target_lang, config
            )
            _write_translation("stderr", translated, sys.stderr, ANSI_YELLOW)
    if _should_translate_stdout(command, result.stdout, target_lang, config):
        stdout_text = _decode_for_translation(result.stdout)
        if stdout_text is not None:
            stream = sys.stdout if sys.stdout.isatty() else sys.stderr
            translated = await _translate_output(
                command, stdout_text, target_lang, config
            )
            _write_translation("stdout", translated, stream, ANSI_CYAN)


async def _translate_output(
    command: list[str], text: str, target_lang: str, config: HotConfig
) -> str:
    cache_command, cache_subcommand = _cache_identity(command)
    return await translate_cached_text(
        cache_command,
        cache_subcommand,
        text,
        target_lang,
        config,
    )


def _cache_identity(command: list[str]) -> tuple[str, str]:
    executable = Path(command[0]).name
    subcommand = command[1] if len(command) > 1 else ""
    return executable, subcommand


def _write_translation(
    stream_name: str, translated: str, stream: TextIO, color: str
) -> None:
    use_color = stream.isatty()
    prefix = f"\n[hymt translated {stream_name}]\n"
    if use_color:
        stream.write(f"{color}{prefix}{translated}{ANSI_RESET}")
    else:
        stream.write(f"{prefix}{translated}")
    if not translated.endswith("\n"):
        stream.write("\n")
    stream.flush()


def _should_translate_stdout(
    command: list[str], output: bytes, target_lang: str, config: HotConfig
) -> bool:
    del target_lang
    if not output:
        return False
    mode = config.exec_translate_stdout
    if mode is False:
        return False
    if _command_matches(command, config.exec_skip_commands):
        return False
    if _matches_skip_pattern(command, config.exec_skip_patterns):
        return False
    if _looks_binary(output):
        return False
    text = _decode_for_translation(output)
    if text is None:
        return False
    if _looks_structured(text) or _looks_like_build_progress(text):
        return False
    return True


def _command_matches(command: list[str], names: Iterable[str]) -> bool:
    executable = Path(command[0]).name
    return executable in set(names)


def _matches_skip_pattern(command: list[str], patterns: Iterable[str]) -> bool:
    if not patterns:
        return False
    candidates = {
        Path(command[0]).name,
        command[0],
        " ".join(command),
    }
    return any(
        fnmatch.fnmatchcase(candidate, pattern)
        for pattern in patterns
        for candidate in candidates
    )


def _decode_for_translation(output: bytes) -> str | None:
    if _looks_binary(output):
        return None
    try:
        return output.decode("utf-8")
    except UnicodeDecodeError:
        return output.decode("utf-8", errors="replace")


def _looks_binary(output: bytes) -> bool:
    sample = output[:4096]
    if b"\x00" in sample:
        return True
    if not sample:
        return False
    allowed_controls = {7, 8, 9, 10, 12, 13, 27}
    control_count = sum(
        1 for byte in sample if byte < 32 and byte not in allowed_controls
    )
    return control_count / len(sample) > 0.05


def _looks_structured(text: str) -> bool:
    stripped = text.strip()
    if not stripped:
        return False
    if _looks_json(stripped) or _looks_xml(stripped):
        return True
    return _looks_yaml(stripped)


def _looks_json(text: str) -> bool:
    if text[0] not in "{[":
        return False
    try:
        json.loads(text)
    except json.JSONDecodeError:
        return False
    return True


def _looks_xml(text: str) -> bool:
    if not text.startswith(("<", "<?xml")):
        return False
    first_line = text.splitlines()[0]
    return ">" in first_line


def _looks_yaml(text: str) -> bool:
    lines = [
        line for line in text.splitlines() if line.strip() and not line.startswith("#")
    ]
    if not lines:
        return False
    if lines[0].strip() in {"---", "..."}:
        return True
    key_value_lines = sum(1 for line in lines[:20] if ":" in line.split("#", 1)[0])
    return len(lines) >= 3 and key_value_lines / min(len(lines), 20) >= 0.8


def _looks_like_build_progress(text: str) -> bool:
    progress_prefixes = (
        "compiling ",
        "checking ",
        "building ",
        "running ",
        "finished ",
        "linking ",
        "generating ",
    )
    lines = [line.strip().lower() for line in text.splitlines() if line.strip()]
    if not lines:
        return False
    if any("error" in line or "warning" in line for line in lines):
        return False
    return all(line.startswith(progress_prefixes) for line in lines)


def is_agent_descendant() -> bool:
    agent_names = frozenset(
        {
            "claude",
            "claude-code",
            "csa",
            "codex",
            "gemini-cli",
            "gemini",
            "opencode",
            "aider",
            "cursor",
            "copilot",
            "antigravity-cli",
        }
    )
    pid = os.getpid()
    while pid > 1:
        try:
            proc = Path(f"/proc/{pid}")
            comm = (proc / "comm").read_text(encoding="utf-8").strip().lower()
            if comm in agent_names:
                return True
            cmdline = (proc / "cmdline").read_bytes()
            cmdline_text = cmdline.replace(b"\x00", b" ").decode(
                "utf-8", errors="replace"
            )
            if any(name in cmdline_text.lower() for name in agent_names):
                return True
            status = (proc / "status").read_text(encoding="utf-8")
            parent = _parent_pid(status)
            if parent is None:
                return False
            pid = parent
        except (OSError, ValueError):
            return False
    return False


def _parent_pid(status: str) -> int | None:
    for line in status.splitlines():
        if line.startswith("PPid:"):
            return int(line.split()[1])
    return None

File: crates/hymt-translate/src/exec_wrapper.rs
//! Run a subprocess, capture its output, translate it, and print the translation.
//!
//! The command is run synchronously; the translation is async.  Binary output,
//! structured data (JSON, XML, YAML), and build-progress lines are skipped.

use std::io::IsTerminal;
use std::path::Path;
use std::process::Command;

use anyhow::Result;

use hymt_cache::history::HistoryDB;
use hymt_cache::ExecCache;
use hymt_client::TranslationClient;
use hymt_core::config::HotConfig;
use hymt_core::language::resolve_target_language;
use hymt_core::templates::{PromptOpts, TemplateType};
use hymt_segment::Segmenter;

use crate::translate::{translate_text, TranslationCtx};

const ANSI_CYAN: &str = "\x1b[36m";
const ANSI_YELLOW: &str = "\x1b[33m";
const ANSI_RESET: &str = "\x1b[0m";

// ── Public entry point ────────────────────────────────────────────────────────

/// Run `command`, capture stdout/stderr, translate the output, and print it.
/// Returns the subprocess exit code.
pub async fn run_exec_command(
    command: &[&str],
    target_lang: &str,
    config: &HotConfig,
    client: &TranslationClient,
    segmenter: &Segmenter,
    history: &HistoryDB,
    explicit_target: bool,
) -> Result<i32> {
    if command.is_empty() {
        anyhow::bail!("command is required");
    }

    let (stdout_bytes, stderr_bytes, exit_code) = run_subprocess(command)?;

    let effective_lang = if explicit_target {
        target_lang.to_owned()
    } else {
        // Auto-detect from stdout sample
        let sample = std::str::from_utf8(&stdout_bytes[..stdout_bytes.len().min(4096)])
            .unwrap_or("")
            .to_owned();
        resolve_target_language(
            &sample,
            target_lang,
            &config.primary_lang(),
            &config.secondary_lang(),
            false,
        )
    };

    let exe = Path::new(command[0])
        .file_name()
        .map(|n| n.to_string_lossy().into_owned())
        .unwrap_or_else(|| command[0].to_owned());
    let subcmd = command.get(1).copied().unwrap_or("").to_owned();

    let tctx = TranslationCtx {
        config,
        client,
        segmenter,
        history,
    };

    // Translate stderr if configured
    if config.exec_translate_stderr() && !stderr_bytes.is_empty() {
        if let Some(text) = decode_for_translation(&stderr_bytes) {
            let cache = ExecCache::new(config.exec_shared_cache_path());
            match translate_cached(&exe, &subcmd, &text, &effective_lang, &cache, &tctx).await {
                Ok(translated) => {
                    write_translation("stderr", &translated, false, ANSI_YELLOW);
                }
                Err(e) => eprintln!("hymt: stderr translation failed: {e}"),
            }
        }
    }

    // Translate stdout if not binary/structured/progress
    let should_translate = should_translate_stdout(command, &stdout_bytes, config);
    if should_translate {
        if let Some(text) = decode_for_translation(&stdout_bytes) {
            let cache = ExecCache::new(config.exec_shared_cache_path());
            let use_tty = std::io::stdout().is_terminal();
            match translate_cached(&exe, &subcmd, &text, &effective_lang, &cache, &tctx).await {
                Ok(translated) => {
                    write_translation("stdout", &translated, use_tty, ANSI_CYAN);
                }
                Err(e) => eprintln!("hymt: stdout translation failed: {e}"),
            }
        }
    }

    Ok(exit_code)
}

// ── Internal helpers ──────────────────────────────────────────────────────────

fn run_subprocess(command: &[&str]) -> Result<(Vec<u8>, Vec<u8>, i32)> {
    let output = Command::new(command[0])
        .args(&command[1..])
        .output()
        .map_err(|e| anyhow::anyhow!("failed to run {:?}: {e}", command[0]))?;

    // Show output to the user in real-time is not possible with Command::output(),
    // so we print it after capture.
    use std::io::Write;
    std::io::stdout().write_all(&output.stdout).ok();
    std::io::stderr().write_all(&output.stderr).ok();

    let code = output.status.code().unwrap_or(1);
    Ok((output.stdout, output.stderr, code))
}

pub(crate) async fn translate_cached(
    command: &str,
    subcommand: &str,
    text: &str,
    target_lang: &str,
    cache: &ExecCache,
    ctx: &TranslationCtx<'_>,
) -> Result<String> {
    if let Ok(Some(cached)) = cache.find(command, subcommand, text, target_lang) {
        return Ok(cached);
    }

    let opts = PromptOpts::default();
    let translated = translate_text(text, target_lang, &TemplateType::Default, &opts, ctx).await?;

    if let Err(e) = cache.store_user(command, subcommand, text, target_lang, &translated) {
        eprintln!("Warning: exec cache store failed: {e}");
    }
    Ok(translated)
}

fn write_translation(stream_name: &str, translated: &str, use_color: bool, color: &str) {
    let prefix = format!("\n[hymt translated {stream_name}]\n");
    if use_color {
        eprint!("{color}{prefix}{translated}{ANSI_RESET}");
    } else {
        eprint!("{prefix}{translated}");
    }
    if !translated.ends_with('\n') {
        eprintln!();
    }
}

fn should_translate_stdout(command: &[&str], output: &[u8], config: &HotConfig) -> bool {
    if output.is_empty() {
        return false;
    }
    if !config.exec_translate_stdout() {
        return false;
    }
    let exe = Path::new(command[0])
        .file_name()
        .map(|n| n.to_string_lossy().into_owned())
        .unwrap_or_default();
    if config.exec_skip_commands().contains(&exe) {
        return false;
    }
    if matches_skip_pattern(command, &config.exec_skip_patterns()) {
        return false;
    }
    if looks_binary(output) {
        return false;
    }
    let text = match decode_for_translation(output) {
        Some(t) => t,
        None => return false,
    };
    if looks_structured(&text) || looks_like_build_progress(&text) {
        return false;
    }
    true
}

fn matches_skip_pattern(command: &[&str], patterns: &[String]) -> bool {
    if patterns.is_empty() {
        return false;
    }
    let exe = Path::new(command[0])
        .file_name()
        .map(|n| n.to_string_lossy().into_owned())
        .unwrap_or_else(|| command[0].to_owned());
    let full = command.join(" ");
    for pattern in patterns {
        for candidate in &[&exe, command[0], &full] {
            if glob_match(candidate, pattern) {
                return true;
            }
        }
    }
    false
}

fn glob_match(text: &str, pattern: &str) -> bool {
    // Simple glob: only * wildcard supported
    let mut t = text;
    let parts: Vec<&str> = pattern.split('*').collect();
    if parts.len() == 1 {
        return t == pattern;
    }
    if !t.starts_with(parts[0]) {
        return false;
    }
    t = &t[parts[0].len()..];
    for (i, part) in parts[1..].iter().enumerate() {
        if i == parts.len() - 2 {
            // Last part must match end
            return t.ends_with(part);
        }
        if let Some(pos) = t.find(part) {
            t = &t[pos + part.len()..];
        } else {
            return false;
        }
    }
    true
}

pub(crate) fn decode_for_translation(output: &[u8]) -> Option<String> {
    if looks_binary(output) {
        return None;
    }
    Some(String::from_utf8_lossy(output).into_owned())
}

// ── Detection functions (pub for tests) ───────────────────────────────────────

/// Returns `true` if `output` appears to be binary data.
///
/// Heuristic: null bytes OR >5% of the first 4 KiB are non-whitespace ASCII
/// control characters.
pub fn looks_binary(output: &[u8]) -> bool {
    let sample = &output[..output.len().min(4096)];
    if sample.contains(&0u8) {
        return true;
    }
    if sample.is_empty() {
        return false;
    }
    const ALLOWED: &[u8] = &[7, 8, 9, 10, 12, 13, 27]; // bell, bs, tab, lf, ff, cr, esc
    let control_count = sample
        .iter()
        .filter(|&&b| b < 32 && !ALLOWED.contains(&b))
        .count();
    control_count * 100 / sample.len() > 5
}

/// Returns `true` if `text` looks like structured data (JSON, XML, or YAML).
pub fn looks_structured(text: &str) -> bool {
    let stripped = text.trim();
    if stripped.is_empty() {
        return false;
    }
    looks_json(stripped) || looks_xml(stripped) || looks_yaml(stripped)
}

fn looks_json(text: &str) -> bool {
    if !matches!(text.chars().next(), Some('{') | Some('[')) {
        return false;
    }
    serde_json::from_str::<serde_json::Value>(text).is_ok()
}

fn looks_xml(text: &str) -> bool {
    if !text.starts_with('<') {
        return false;
    }
    let first_line = text.lines().next().unwrap_or("");
    first_line.contains('>')
}

fn looks_yaml(text: &str) -> bool {
    let lines: Vec<&str> = text
        .lines()
        .filter(|l| !l.trim().is_empty() && !l.starts_with('#'))
        .collect();
    if lines.is_empty() {
        return false;
    }
    if lines[0].trim() == "---" || lines[0].trim() == "..." {
        return true;
    }
    let kv = lines[..lines.len().min(20)]
        .iter()
        .filter(|l| l.split('#').next().unwrap_or("").contains(':'))
        .count();
    let sample_size = lines.len().min(20);
    sample_size >= 3 && kv * 10 >= sample_size * 8 // ≥80%
}

/// Returns `true` if every non-empty line starts with a build-progress verb.
pub fn looks_like_build_progress(text: &str) -> bool {
    const PREFIXES: &[&str] = &[
        "compiling ",
        "checking ",
        "building ",
        "running ",
        "finished ",
        "linking ",
        "generating ",
    ];
    let lines: Vec<&str> = text
        .lines()
        .map(|l| l.trim())
        .filter(|l| !l.is_empty())
        .collect();
    if lines.is_empty() {
        return false;
    }
    // If any line has an error/warning keyword, not build progress
    if lines
        .iter()
        .any(|l| l.contains("error") || l.contains("warning"))
    {
        return false;
    }
    lines
        .iter()
        .all(|l| PREFIXES.iter().any(|p| l.to_lowercase().starts_with(p)))
}

// ── Agent-descendant detection (Linux-specific) ───────────────────────────────

/// Returns `true` if the process tree contains a known AI-agent process.
#[cfg(target_os = "linux")]
pub fn is_agent_descendant() -> bool {
    const AGENTS: &[&str] = &[
        "claude",
        "claude-code",
        "csa",
        "codex",
        "gemini-cli",
        "gemini",
        "opencode",
        "aider",
        "cursor",
        "copilot",
    ];
    let mut pid = std::process::id();
    loop {
        if pid <= 1 {
            return false;
        }
        let comm_path = format!("/proc/{pid}/comm");
        if let Ok(comm) = std::fs::read_to_string(&comm_path) {
            let comm = comm.trim().to_lowercase();
            if AGENTS.iter().any(|&a| comm == a) {
                return true;
            }
        }
        // Read parent PID from /proc/{pid}/status
        let status_path = format!("/proc/{pid}/status");
        let status = match std::fs::read_to_string(&status_path) {
            Ok(s) => s,
            Err(_) => return false,
        };
        let ppid = status
            .lines()
            .find(|l| l.starts_with("PPid:"))
            .and_then(|l| l.split_whitespace().nth(1))
            .and_then(|s| s.parse::<u32>().ok());
        match ppid {
            Some(p) if p > 0 => pid = p,
            _ => return false,
        }
    }
}

#[cfg(not(target_os = "linux"))]
pub fn is_agent_descendant() -> bool {
    false
}

// ── Tests ─────────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    // ── looks_binary ─────────────────────────────────────────────────────────

    #[test]
    fn binary_null_byte_is_binary() {
        assert!(looks_binary(b"hello\x00world"));
    }

    #[test]
    fn binary_empty_is_not_binary() {
        assert!(!looks_binary(b""));
    }

    #[test]
    fn binary_plain_text_is_not_binary() {
        assert!(!looks_binary(b"Hello, world!\nThis is normal text.\n"));
    }

    #[test]
    fn binary_high_control_ratio_is_binary() {
        // 10 control bytes among 20 bytes = 50% > 5%
        let data: Vec<u8> = (0u8..20).collect();
        assert!(looks_binary(&data));
    }

    #[test]
    fn binary_ansi_escape_is_not_binary() {
        // ESC (27) is in the allowed set
        let data = b"\x1b[0mHello\x1b[1mWorld\x1b[0m";
        assert!(!looks_binary(data));
    }

    // ── looks_structured ─────────────────────────────────────────────────────

    #[test]
    fn structured_valid_json_object() {
        assert!(looks_structured(r#"{"key": "value"}"#));
    }

    #[test]
    fn structured_valid_json_array() {
        assert!(looks_structured(r#"[1, 2, 3]"#));
    }

    #[test]
    fn structured_invalid_json_not_structured() {
        assert!(!looks_structured("{not valid json}"));
    }

    #[test]
    fn structured_xml_first_line_has_tag() {
        assert!(looks_structured("<root>\n<child/>\n</root>"));
    }

    #[test]
    fn structured_yaml_document_marker() {
        assert!(looks_structured("---\nkey: value\n"));
    }

    #[test]
    fn structured_high_kv_ratio_yaml() {
        let yaml = "key1: val1\nkey2: val2\nkey3: val3\n";
        assert!(looks_structured(yaml));
    }

    #[test]
    fn structured_plain_text_not_structured() {
        assert!(!looks_structured(
            "This is a normal paragraph of text with no structure."
        ));
    }

    // ── looks_like_build_progress ─────────────────────────────────────────────

    #[test]
    fn build_progress_all_progress_lines() {
        let text = "Compiling foo v1.0\nChecking bar v2.0\nFinished dev\n";
        assert!(looks_like_build_progress(text));
    }

    #[test]
    fn build_progress_with_error_is_not_progress() {
        let text = "Compiling foo v1.0\nerror[E0001]: something\n";
        assert!(!looks_like_build_progress(text));
    }

    #[test]
    fn build_progress_mixed_content_is_not_progress() {
        assert!(!looks_like_build_progress("Hello, this is output.\n"));
    }

    #[test]
    fn build_progress_empty_is_not_progress() {
        assert!(!looks_like_build_progress(""));
    }
}
